File: desktop/src/types/message.ts
import { User } from "./user";

export interface ReactionGroup {
  emoji: string;
  count: number;
  users: string[];
}

export interface FileAttachment {
  id: string;
  message_id?: string;
  channel_id: string;
  user_id: string;
  filename: string;
  content_type: string;
  size_bytes: number;
  download_url?: string;
  created_at: string;
}

export interface Message {
  id: string;
  channel_id: string;
  user_id: string;
  parent_id?: string;
  content: string;
  is_alert: boolean;
  alert_severity?: "info" | "warning" | "critical";
  alert_metadata?: Record<string, unknown>;
  edited_at?: string;
  deleted_at?: string;
  created_at: string;
  user?: User;
  reactions?: ReactionGroup[];
  attachments?: FileAttachment[];
  reply_count: number;
}

export interface CreateMessageRequest {
  content: string;
  parent_id?: string;
  attachment_ids?: string[];
}

export interface UpdateMessageRequest {
  content: string;
}

File: server/internal/model/message.go
package model

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

type Message struct {
	ID            uuid.UUID        `json:"id"`
	ChannelID     uuid.UUID        `json:"channel_id"`
	UserID        uuid.UUID        `json:"user_id"`
	ParentID      *uuid.UUID       `json:"parent_id,omitempty"`
	Content       string           `json:"content"`
	IsAlert       bool             `json:"is_alert"`
	AlertSeverity *string          `json:"alert_severity,omitempty"`
	AlertMetadata json.RawMessage  `json:"alert_metadata,omitempty"`
	EditedAt      *time.Time       `json:"edited_at,omitempty"`
	DeletedAt     *time.Time       `json:"deleted_at,omitempty"`
	CreatedAt     time.Time        `json:"created_at"`
	User          *User            `json:"user,omitempty"`
	Reactions     []ReactionGroup  `json:"reactions,omitempty"`
	Attachments   []FileAttachment `json:"attachments,omitempty"`
	ReplyCount    int              `json:"reply_count"`
}

type CreateMessageRequest struct {
	Content       string      `json:"content" validate:"required,min=1,max=10000"`
	ParentID      *uuid.UUID  `json:"parent_id"`
	AttachmentIDs []uuid.UUID `json:"attachment_ids"`
}

type UpdateMessageRequest struct {
	Content string `json:"content" validate:"required,min=1,max=10000"`
}

type MessageListParams struct {
	ChannelID uuid.UUID
	Before    *uuid.UUID
	Limit     int
}
